sycl/plugins/esimd_cpu/esimd_libcm.hpp

```diff
@@ -0,0 +1,327 @@
+
+#ifndef __SYCL_DEVICE_ONLY__
+
+#include <CL/sycl/backend_types.hpp>
+#include <CL/sycl/detail/accessor_impl.hpp>
+#include <CL/sycl/detail/common.hpp>
+#include <CL/sycl/detail/export.hpp>
+#include <CL/sycl/detail/helpers.hpp>
+#include <CL/sycl/detail/host_profiling_info.hpp>
+#include <CL/sycl/detail/kernel_desc.hpp>
+#include <CL/sycl/detail/type_traits.hpp>
+#include <CL/sycl/group.hpp>
+#include <CL/sycl/id.hpp>
+#include <CL/sycl/kernel.hpp>
+#include <CL/sycl/nd_item.hpp>
+#include <CL/sycl/range.hpp>
+
+#include <esimdcpu_runtime.h>
+
+#include <functional>
+#include <thread>
+#include <vector>
+
+#define LAMBDA_WRAPPER_TMPL(ARGTYPE, TAG)                                      \
+  typedef std::function<void(const ARGTYPE &)> LambdaFunction_##TAG;           \
+                                                                               \
+  extern "C" struct LambdaWrapper_##TAG {                                      \
+    LambdaFunction_##TAG Func;                                                 \
+    std::vector<ARGTYPE> ArgVector;                                            \
+  };                                                                           \
+                                                                               \
+  template <typename LambdaTy>                                                 \
+  auto makeWrapper_##TAG(std::vector<ARGTYPE> &ArgVector, LambdaTy F) {        \
+    std::unique_ptr<LambdaWrapper_##TAG> Wrapper(new LambdaWrapper_##TAG);     \
+    Wrapper->Func = LambdaFunction_##TAG(F);                                   \
+    Wrapper->ArgVector = ArgVector;                                            \
+    return Wrapper;                                                            \
+  }                                                                            \
+                                                                               \
+  extern "C" inline void invokeLambda_##TAG(void *Wrapper) {                   \
+    auto *w = reinterpret_cast<LambdaWrapper_##TAG *>(Wrapper);                \
+    w->Func(w->ArgVector[cm_support::thread_local_idx()]);                     \
+  }
+
+#define _COMMA_ ,
+
+LAMBDA_WRAPPER_TMPL(sycl::id<3>, ID_3DIM)
+LAMBDA_WRAPPER_TMPL(sycl::id<2>, ID_2DIM)
+LAMBDA_WRAPPER_TMPL(sycl::id<1>, ID_1DIM)
+LAMBDA_WRAPPER_TMPL(sycl::item<3 _COMMA_ false>, ITEM_3DIM)
+LAMBDA_WRAPPER_TMPL(sycl::item<2 _COMMA_ false>, ITEM_2DIM)
+LAMBDA_WRAPPER_TMPL(sycl::item<1 _COMMA_ false>, ITEM_1DIM)
+LAMBDA_WRAPPER_TMPL(sycl::item<3 _COMMA_ true>, ITEM_OFFSET_3DIM)
+LAMBDA_WRAPPER_TMPL(sycl::item<2 _COMMA_ true>, ITEM_OFFSET_2DIM)
+LAMBDA_WRAPPER_TMPL(sycl::item<1 _COMMA_ true>, ITEM_OFFSET_1DIM)
+LAMBDA_WRAPPER_TMPL(sycl::nd_item<3>, NDITEM_3DIM)
+LAMBDA_WRAPPER_TMPL(sycl::nd_item<2>, NDITEM_2DIM)
+LAMBDA_WRAPPER_TMPL(sycl::nd_item<1>, NDITEM_1DIM)
+
+#undef LAMBDA_WRAPPER_TMPL
+#undef _COMMA_
+
+template <class KernelType, class KernelArgType, int __Dims__>
```

Comment:
please add comments - for the class and the fields

Reply:
Will do.